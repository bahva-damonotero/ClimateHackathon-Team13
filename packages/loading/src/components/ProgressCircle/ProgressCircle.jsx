import React from 'react';
import PropTypes from 'prop-types';

import { describeArc } from './utils';

export default function ProgressCircle({
  color,
  fontFamily,
  fontSize,
  fontWeight,
  rounded,
  showPercentage,
  size,
  strokeWidth,
  value
}) {
  const actualStrokeWidth = strokeWidth || size / 26;
  const radius = (size - actualStrokeWidth) / 2;
  const center = radius + actualStrokeWidth / 2;
  const endAngle = ((value >= 100 ? 99.9999 : value) / 100) * 360;

  return (
    <div
      aria-live="assertive"
      className="bah-uswds-progress-circle"
      role="alert"
      style={{ width: size, height: size }}
    >
      <svg aria-hidden="true" height={size} width={size}>
        <path
          d={describeArc(center, center, radius, 0, endAngle)}
          fill="none"
          stroke={color}
          strokeWidth={actualStrokeWidth}
        />
        {showPercentage && (
          <text
            x={center}
            y={center - size / 70}
            fontFamily={fontFamily}
            fontSize={fontSize || size / 4}
            fontWeight={fontWeight}
            dominantBaseline="central"
            textAnchor="middle"
          >
            {rounded ? Math.floor(value) : value}%
          </text>
        )}
      </svg>

      <div className="bah-uswds-loading--sr-text">Progress:{value}%</div>
    </div>
  );
}

ProgressCircle.propTypes = {
  /** Color of the progress circle */
  color: PropTypes.string,
  /** Font family used to display percentage */
  fontFamily: PropTypes.string,
  /** Font size used to display percentage */
  fontSize: PropTypes.number,
  /** Font weight used to display percentage */
  fontWeight: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf(['normal', 'bold', 'bolder', 'lighter'])
  ]),
  /** If `false`, the displayed percentage will not be rounded down */
  rounded: PropTypes.bool,
  /** If `false`, the percentage is not displayed */
  showPercentage: PropTypes.bool,
  /** Size of the progress circle in pixels */
  size: PropTypes.number,
  /** Stroke width of the progress circle in pixels */
  strokeWidth: PropTypes.number,
  /** Percentage (out of 100) of the progress circle */
  value: PropTypes.number.isRequired
};

ProgressCircle.defaultProps = {
  color: 'blue',
  rounded: true,
  showPercentage: true,
  size: 80
};
