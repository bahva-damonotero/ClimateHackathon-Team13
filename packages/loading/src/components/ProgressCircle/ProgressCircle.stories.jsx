import React from 'react';

import ProgressCircle from '.';

export default {
  title: 'Loading Indicators/ProgressCircle',
  component: ProgressCircle
};

const Template = (args) => <ProgressCircle {...args} />;

export const differentColor = Template.bind({});
differentColor.args = {
  color: 'red',
  value: 100
};

export const noPercentage = Template.bind({});
noPercentage.args = {
  showPercentage: false,
  value: 75.5
};

export const noRounding = Template.bind({});
noRounding.args = {
  rounded: false,
  value: 75.5
};

export const differentStrokeWidth = Template.bind({});
differentStrokeWidth.args = {
  strokeWidth: 5,
  value: 75.5
};

export const fontOverrides = Template.bind({});
fontOverrides.args = {
  fontFamily: 'Verdana',
  fontSize: 12,
  fontWeight: 'bold',
  value: 45
};
