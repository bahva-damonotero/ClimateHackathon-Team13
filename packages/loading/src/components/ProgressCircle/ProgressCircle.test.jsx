import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import ProgressCircle from '.';

describe('ProgressCircle', () => {
  const color = 'blue';
  const fontFamily = '';
  const fontWeight = 'normal';
  const size = 80;
  const strokeWidth = 10;
  const value = 10.5;
  const title = `Progress:${value}%`;

  const PROGRESS_SELECTOR =
    '.bah-uswds-progress-circle > .bah-uswds-loading--sr-text';
  const CIRCLE_SELECTOR = '.bah-uswds-progress-circle';

  const jsx = (
    <ProgressCircle
      color={color}
      fontFamily={fontFamily}
      fontWeight={fontWeight}
      size={size}
      strokeWidth={strokeWidth}
      value={value}
    />
  );

  function getProgressTitle(container) {
    return container.querySelector(PROGRESS_SELECTOR);
  }

  function getProgressCircle(container) {
    return container.querySelector(CIRCLE_SELECTOR);
  }

  it('should render properly', () => {
    const { container } = render(jsx);

    expect(getProgressCircle(container).children.length).toBe(2);
    expect(getProgressTitle(container).innerHTML).toBe(title);
    expect(
      getProgressCircle(container).querySelector('.bah-uswds-loading--sr-text')
    ).toBeTruthy();
    expect(getProgressCircle(container).getAttribute('aria-live')).toBe(
      'assertive'
    );
    expect(getProgressCircle(container).getAttribute('role')).toBe('alert');
    expect(getProgressCircle(container).getAttribute('style')).toBe(
      'width: 80px; height: 80px;'
    );
    expect(container.querySelector('text').getAttribute('y')).toBe(
      '38.857142857142854'
    );
    expect(container.querySelector('text').getAttribute('x')).toBe('40');
    expect(container.querySelector('path').getAttribute('stroke')).toBe('blue');
  });

  it('should render a decimal number when rounded is false', () => {
    const { container } = render(
      <ProgressCircle
        color={color}
        fontFamily={fontFamily}
        fontWeight={fontWeight}
        rounded={false}
        size={size}
        strokeWidth={strokeWidth}
        value={value}
      />
    );

    expect(getProgressTitle(container).innerHTML).toBe(title);
  });

  it('should render a rounded number when rounded variable is true', () => {
    const { container } = render(jsx);

    expect(container.querySelector('text').innerHTML).toBe('10%');
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
