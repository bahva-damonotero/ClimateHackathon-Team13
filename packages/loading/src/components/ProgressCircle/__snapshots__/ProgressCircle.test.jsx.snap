// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressCircle should match snapshot 1`] = `
<div
  aria-live="assertive"
  className="bah-uswds-progress-circle"
  role="alert"
  style={
    Object {
      "height": 80,
      "width": 80,
    }
  }
>
  <svg
    aria-hidden="true"
    height={80}
    width={80}
  >
    <path
      d="M 61.45174687785418 12.344574566850834 A 35 35 0 0 0 40 5"
      fill="none"
      stroke="blue"
      strokeWidth={10}
    />
    <text
      dominantBaseline="central"
      fontFamily=""
      fontSize={20}
      fontWeight="normal"
      textAnchor="middle"
      x={40}
      y={38.857142857142854}
    >
      10
      %
    </text>
  </svg>
  <div
    className="bah-uswds-loading--sr-text"
  >
    Progress:
    10.5
    %
  </div>
</div>
`;
